sensor/kubernetes/listener/resources/hash_reconciliation.go

```diff
@@ -0,0 +1,75 @@
+package resources
+
+import (
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/stringutils"
+)
+
+// ResourceStoreReconciler handles sensor-side reconciliation using in-memory store
+type ResourceStoreReconciler struct {
+	storeProvider *InMemoryStoreProvider
+}
+
+// NewResourceStoreReconciler builds ResourceStoreReconciler for sensor-side reconciliation
+func NewResourceStoreReconciler(storeProvider *InMemoryStoreProvider) *ResourceStoreReconciler {
+	return &ResourceStoreReconciler{storeProvider: storeProvider}
+}
+
+// ProcessHashes orchestrates the sensor-side reconciliation after a reconnect. It returns a slice of resource IDs that
+// should be deleted in Central to keep the state of Sensor and Central in sync.
+func (hr *ResourceStoreReconciler) ProcessHashes(h map[string]uint64) []central.MsgFromSensor {
+	events := make([]central.MsgFromSensor, 0)
+	for typeWithID, hashValue := range h {
+		resType, resID := stringutils.Split2(typeWithID, ":")
+		if resID == "" {
+			log.Errorf("malformed hash key: %s", typeWithID)
+			continue
+		}
+		toDeleteID, err := hr.storeProvider.ReconcileDelete(resType, resID, hashValue)
+		if err != nil {
+			log.Errorf("reconciliation error: %s", err)
+		}
+		if toDeleteID == "" {
+			log.Debug("empty reconciliation result - not found on Sensor")
+			continue
+		}
+		delMsg, err := resourceToMessage(resType, toDeleteID)
+		if err != nil {
+			log.Errorf("converting resource to MsgFromSensor: %s", err)
+			continue
+		}
+		events = append(events, *delMsg)
+	}
+	return events
+}
+
+func resourceToMessage(resType string, resID string) (*central.MsgFromSensor, error) {
```

Comment:
I couldn't find a better way to make this generic. It would be easy if the protobuf generated code would make `isSensorEvent_Resource` public, so then we could leverage it to generate `central.SensorEvent` dynamically. 😞